pkg/agent/controller/traceflow/traceflow_controller.go

```diff
@@ -0,0 +1,386 @@
+func (c *Controller) processTraceflowItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	// We call Done here so the workqueue knows we have finished processing this item. We also
+	// must remember to call Forget if we do not want this work item being re-queued. For
+	// example, we do not call Forget if a transient error occurs, instead the item is put back
+	// on the workqueue and attempted again after a back-off period.
+	defer c.queue.Done(obj)
+
+	// We expect strings (Traceflow name) to come off the workqueue.
+	if key, ok := obj.(string); !ok {
+		// As the item in the workqueue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen: enqueueTraceflow only enqueues strings.
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncTraceflow(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again.
+		c.queue.Forget(key)
+	} else {
+		// If error occurs we log error.
+		klog.Errorf("Error syncing Traceflow %s, Aborting. Error: %v", key, err)
+	}
+	return true
+}
+
+// syncTraceflow gets Traceflow CRD by name, update cache and start syncing.
+func (c *Controller) syncTraceflow(traceflowName string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing Traceflow for %s. (%v)", traceflowName, time.Since(startTime))
+	}()
+
+	tf, err := c.traceflowLister.Get(traceflowName)
+	if err != nil {
+		return err
+	}
+	switch tf.Status.Phase {
+	case opsv1alpha1.Running:
+		if tf.Status.DataplaneTag != 0 {
+			if _, ok := c.running[tf.Status.DataplaneTag]; !ok {
+				c.running[tf.Status.DataplaneTag] = *tf
+				err = c.startTraceflow(tf)
+			} else {
+				c.UpdateTagCache(tf)
+			}
+		} else {
+			klog.Warningf("Invalid data plane tag %d for Traceflow %s", tf.Status.DataplaneTag, tf.Name)
+		}
+	default:
+		c.deallocateTag(tf)
+	}
+	return err
+}
+
+// startTraceflow deploys OVS flow entries for Traceflow and inject packet if current Node
+// is Sender Node.
+func (c *Controller) startTraceflow(tf *opsv1alpha1.Traceflow) error {
+	// Deploy flow entries for traceflow
+	klog.V(2).Infof("Deploy flow entries for Traceflow %s", tf.Name)
+	err := c.ofClient.InstallTraceflowFlows(tf.Status.DataplaneTag)
+	defer func() {
+		if err != nil {
+			c.errorTraceflowCRD(tf)
+		}
+	}()
+	if err != nil {
+		return err
+	}
+
+	// Inject packet if this Node is sender.
+	time.Sleep(time.Duration(injectPacketDelay) * time.Second)
+	err = c.injectPacket(tf)
+	return err
+}
+
+func (c *Controller) injectPacket(tf *opsv1alpha1.Traceflow) error {
+	namespaceName := tf.Spec.Source.Namespace
+	podName := tf.Spec.Source.Pod
+	podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(podName, namespaceName)
+	// Pod not found in current Node, skip inject packet.
```

Comment:
Here it is possible the Pod is on the Node, but Pod interface is not created (or failed). Maybe we should let controller compute the source Node, and the source Node can just return an error, if fails to find the Pod.
Of course I do not mean this PR. But maybe you can add a TODO here.

Reply:
Added TODO